librad/src/identities/generic/tests.rs

```diff
@@ -0,0 +1,201 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use std::collections::{BTreeMap, BTreeSet};
+
+use nonempty::NonEmpty;
+use proptest::prelude::*;
+
+use super::{gen::*, *};
+use crate::{identities::delegation, keys::tests::gen_secret_key};
+
+proptest! {
+    #[test]
+    fn signed_empty_delegations(
+        signing_keys in prop::collection::vec(gen_secret_key(), 1..3).no_shrink(),
+    ) {
+        let signatures = signing_keys
+            .into_iter()
+            .map(|key| (key.public(), key.sign(Boring.as_ref())))
+            .collect::<BTreeMap<_, _>>()
+            .into();
+
+        assert!(matches!(
+            Verifying::from(boring(
+                delegation::Direct::from(BTreeSet::new()),
+                signatures
+            ))
+            .signed::<!>(),
+            Err(error::Verify::NoValidSignatures(_, _))
+        ))
+    }
+
+    #[test]
+    fn signed(id in gen_identity::<Boring>()) {
+        assert_eq!(
+            Verifying::from(id.clone())
+                .signed::<!>()
+                .unwrap()
+                .into_inner(),
+            id
+        )
+    }
+
+    #[test]
+    fn quorum_below_threshold(id in gen_identity::<Boring>()) {
+        // Hitting the threshold exactly is one too few
+        let threshold = id.quorum_threshold();
+        let signatures: Signatures = BTreeMap::from(id.signatures)
+            .into_iter()
+            .take(threshold)
```

Comment:
Is there any way we can test `1..threshold` for this? Technically we're only ever testing `signatures == threshold` but not `< threshold`, bar the `0`-case above.

Reply:
Hm good point, let me `prop_flat_map` something

Reply:
PTAL: d72c71c5a6e6723615b4862858d8e5e8b28ed7b5